indigo/indigo/src/main/scala/indigo/platform/events/GlobalEventStream.scala

```diff
@@ -48,23 +48,52 @@ final class GlobalEventStream(
 
     // Storage
     case StorageEvent.FetchKeyAt(index) =>
-      eventQueue.enqueue(StorageEvent.KeyFoundAt(index, storage.key(index)))
+      storage.key(index) match {
+        case Right(err) => eventQueue.enqueue(err)
+        case Left(str)  => eventQueue.enqueue(StorageEvent.KeyFoundAt(index, str))
+      }
 
     case StorageEvent.FetchKeys(f, t) =>
       val keys = (f to t).toList.map(i => i -> storage.key(i))
-      eventQueue.enqueue(StorageEvent.KeysFound(keys))
+      val errors = keys.flatMap {
+        _ match {
+          case (_, Right(err)) => Some(err)
+          case _               => None
+        }
+      }
+
+      if errors.length > 0 then errors foreach (e => eventQueue.enqueue(e))
+      else
+        eventQueue.enqueue(StorageEvent.KeysFound(keys.flatMap {
+          _ match {
+            case (i, Left(str)) => Some((i, str))
+            case _              => None
+          }
```

Comment:
It's early, but I think you're using some of Scala's implicit conversions here to automagically turn an `Option` into a `List`. If so I'd prefer:

```scala
            case (i, Left(str)) => List((i, str))
            case _              => Nil
```

I really don't think Scala should let you treat Option as List, it causes much confusion, but here we are!

Reply:
That wil give me a `List[List[(Int, String)]]` won't it? Above I'm using `flatMap` to take a `List[(Int, Option[String])]` and convert it to a `List[(Int, String)`. Happy to change it though if I'm wrong :)

Reply:
We're saying that same thing. I'm saying using flatMap to convert `List[Option[A]]` to `List[A]` is totally allowed by Scala, but I consider it bad practice (because I don't think it should be allowed). Don't worry about it now.